### PR TITLE
Save OAuthClient credentials in separate model

### PR DESCRIPTION
**Pull request by wielinde**
- Adds a separate model `OAuthClient` for saving `client_id` and `client_secret` to OP core so that it becomes reusable for other future integrations that need OAuth2.
- The relation between such a `OAuthClient` and some integrating module/feature is polymorphic.
- This PR keeps all controllers and views for adding a `OAuthClient` close to the code that actually is the integration. That way we can have meaningful routes and integration specific labels that provide support and orientation. In the current case that is the `Storages` module.

https://community.openproject.org/projects/openproject/work_packages/42577

### CODE REVIEW

File: app/contracts/oauth_clients/concerns/manage_storages_guarded.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module OAuthClients
+  module Concerns
+    module ManageStoragesGuarded
```
Comment (Kharonus) [inline app/contracts/oauth_clients/concerns/manage_storages_guarded.rb]: maybe that name is misleading? No storage permission checked in here at all. It just checks for the user to be an active admin.

File: app/contracts/oauth_clients/create_contract.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module OAuthClients
+  class CreateContract < ::ModelContract
+    include ActiveModel::Validations
+    include Concerns::ManageStoragesGuarded
+
+    include ::OAuthClients::Concerns::ManageStoragesGuarded
```
Comment (Kharonus) [inline app/contracts/oauth_clients/create_contract.rb]: same include as in line 32. usually I prefer the more specific namespace definition, that starts with root.

File: modules/storages/app/controllers/storages/admin/storages_controller.rb [hunk 3]
```diff
@@ -165,6 +180,10 @@ def ensure_storages_module_active
   def permitted_storage_params
     params
       .require(:storages_storage)
```
Comment (Kharonus) [inline modules/storages/app/controllers/storages/admin/storages_controller.rb:182]: why is this such a weird double named symbol?

File: modules/storages/app/models/storages/storage.rb [hunk 4]
```diff
@@ -51,6 +51,8 @@ class Storages::Storage < ApplicationRecord
   has_many :projects_storages, dependent: :destroy, class_name: 'Storages::ProjectStorage'
   # We can get the list of projects with this Storage enabled.
   has_many :projects, through: :projects_storages
+  # A storage authenticates at a storage via OAuth.
```
Comment (Kharonus) [inline modules/storages/app/models/storages/storage.rb]: confusing comment: more like `A storage authenticates at its remote origin via OAuth.` You can elaborate even a bit more, to make clear, what is "our" storage and what is "the remote storage".

File: app/models/oauth_client.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class OAuthClient < ApplicationRecord
+  belongs_to :integration, polymorphic: true
```
Comment (Kharonus) [inline app/models/oauth_client.rb:30]: is this relation optional?

**Review by Kharonus — CHANGES_REQUESTED** [review]
a few change requests and some questions.

Looks good overall, I did not yet run a manual test. Do you think, at this point of time, a QA of the page from my perspective is making any sense? Or should we do it a bit later, when more parts of the integration are functional?

File: app/contracts/oauth_clients/create_contract.rb (hunk 2 above)
Comment (wielinde) [reply app/contracts/oauth_clients/create_contract.rb]: Well spotted. Tx.

File: app/contracts/oauth_clients/concerns/manage_storages_guarded.rb (hunk 1 above)
Comment (wielinde) [reply app/contracts/oauth_clients/concerns/manage_storages_guarded.rb]: You are right.

File: modules/storages/app/controllers/storages/admin/storages_controller.rb (hunk 3 above)
Comment (wielinde) [reply modules/storages/app/controllers/storages/admin/storages_controller.rb:182]: I believe that is due to the name spacing in the routes.

**Review by Kharonus — APPROVED** [review]
code LGTM

**Pull request merged**